crates/term_input_macros/src/lib.rs

```diff
@@ -13,10 +17,31 @@ pub fn byte_seq_parser(input: TokenStream) -> TokenStream {
         fn_return_type,
         rules,
     } = syn::parse_macro_input!(input as syntax::Input);
-    let fn_body = tree::build_decision_tree(rules);
 
-    quote!(fn #fn_name(buf: &[u8]) -> Option<(#fn_return_type, usize)> {
+    let is_valid_key = build_is_valid_key(&rules);
+
+    let fn_body = tree::build_decision_tree(rules);
+    let byte_parser = quote!(fn #fn_name(buf: &[u8]) -> Option<(#fn_return_type, usize)> {
         #fn_body
-    })
-    .into()
+    });
+
+    quote!(#is_valid_key #byte_parser).into()
+}
+
+fn build_is_valid_key(rules: &[Rule]) -> TokenStream2 {
+    let keys = rules
+        .iter()
+        .map(|r| &r.value.0)
+        .collect::<HashSet<_>>()
+        .into_iter();
+
+    quote!(
+        pub fn is_valid_key(k: Key) -> bool {
```

Comment:
I think name of this function should be derived from `fn_name` above, otherwise we can't use this macro multiple times in a module.